cctbx/cctbx_project
# Python 3.6 / 2.7 cross-compatibility

**This pull request implements the migration of cctbx_project to Python 3.6 while retaining cross-compatibility with Python 2.7.   All tests within cctbx_project were tested on Centos-6.  Mac OSX 10.11 works, and a few more tests fail on 10.14.  Windows tests are still in the future.  The current work was limited to assuring that the tests pass.  We plan follow-up work to assure that all command-line programs run correctly.**

### Current test results are:

|Test results		|	N_total	|	N_fail	|	N_fail	|	N_fail|
|---------------------|-----------------|----------------|----------------|------------------------|
|                               |                       |(Py3.6, Linux)|	(Py2.7, Linux) |(Py27, master, Linux)|
|boost_adaptbx	|	9	|	1	|	0	|	0	|	
|cctbx			|	138		|4	|	0|		0|
|cma_es		|	1	|	0	|	0	|	0|
|fable		|		9	|	0	|	0	|	0|
|gltbx			|	1		|1	|	0	|	0|
|iotbx			|	107		|4	|	0	|	0|
|libtbx			|	36		|0	|	0	|	0|
|mmtbx			|	369	|	112	|	2	|	2|
|rstbx			|	6	|	0	|	0	|	0|
|scitbx			|	135	|	0	|	0	|	0|
|simtbx			|	3	|	0	|	0	|	0|
|smtbx			|	35	|	3	|	0	|	0|
|spotfinder		|	3	|	0	|	0	|	0|
|TOTAL			|	856	|	125	|	2	|	2|
|||||
|without mmtbx	|	487	|	13	|	0	|	0|


### With regard to certain tests:
1. We specifically ask that developers wait for @phyy-nx to fix the boost_adaptbx test (tst_python_streambuf.py), which will be done after the Py2-> Py3 migration is performed for dxtbx (which is intended to be a future community effort).
2. For smtbx, could @luc-j-bourhis and @pcxod have a look?
3. For iota and prime, could @alyubimov please have a look?
4. For mmtbx, @bkpoon, @nwmoriarty, and @olegsobolev will pursue all the follow up on porting tests to Py3.
5. We’ll be requesting help from @Oeffner for Windows issues.
6. The Berkeley group is currently fixing the remaining tests but this is still a short term work in progress.

### To test this branch, use these commands:

Instructions (use two different working directories for Py2 and Py3):
1. Download sources:	`python bootstrap.py hot update --builder=cctbx`
2. Checkout branch:	`cd modules/cctbx_project; git checkout py3`
2a. Replace the downloaded `bootstrap.py` with a symbolic link to the one in `cctbx_project`:
`cd ../..; rm bootstrap.py; ln -s modules/cctbx_project/libtbx/auto_build/bootstrap.py`
3. Build (deactivate any active conda environment, first)
3a. Python 3 Build:	`python bootstrap.py base build --builder=cctbx --use-conda --python3 --nproc=32`
3b. Python 2 Build:	`python bootstrap.py base build --builder=cctbx --use-conda --nproc=32`
4. Source environment:	`source ./build/setpaths.sh`
5. Run tests:		`python bootstrap.py tests --builder=cctbx --nproc=32`

Note, to check for Python 3 syntax: `libtbx.py_compile_all -v -f` with a Python 3 build

### Specific outcomes:
1. We had to switch to the Boost.Python.numpy interface (dropping the numeric interface) to get things working on Python 3. We will be making additional changes to get support for more recent versions of Boost.
2. Because str in Python 3 is unicode text, flex.std_string now returns unicode text in Python 3.  In Python 2, str is ascii text (byte string), so flex.std_string returns ascii text. This simplifies the migration because we do not need to put “b” or “u” in front of every string in our code. 
3. Pickling of flex objects requires byte strings to save and recover state.  To this end, we ensured that the getstate methods of flex objects return byte strings, as do flex.int.copy_to_byte_str() and flex.int_from_byte_str().
4. For cbflib, we had to convert to returning byte strings instead of unicode strings.  See 7bc6def31dc6952d01f3fe9916a5927ca8788a22.
5. Python3 syntax changes were made in the master branch for the following repos:  annlib_adaptbx, ccp4io_adaptbx.

We'd like to thank all the many contributors to this revision, including @nksauter, @bkpoon, @phyy-nx, @dermen, @graeme-winter, @Anthchirp, @ndevenish, @Oeffner.
